File: apps/android/app/src/main/java/com/voiid/app/onboarding/SignupScreen.kt
package com.voiid.app.onboarding

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.material.icons.filled.AlternateEmail
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import com.voiid.app.ui.theme.VoiidSpacing
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.PhotoCamera
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.foundation.Image
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.voiid.app.model.AppSession
import com.voiid.app.net.ProfileService
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import com.voiid.app.ui.theme.VoiidRadius
import java.io.ByteArrayOutputStream

/** Username availability lifecycle, checked against the server as the user types. */
private sealed interface UStatus {
    data object Idle : UStatus
    data object Checking : UStatus
    data object Available : UStatus
    data class Taken(val reason: String) : UStatus
}

/**
 * Onboarding identity step — PHOTO + NAME + USERNAME, port of iOS `SignupScreen`.
 *
 * Field ownership matches iOS: everything REQUIRED about the person is collected here (the
 * verified phone is shown read-only below), and the NEXT screen collects only OPTIONAL extras
 * (email, bio). The single server write happens there, once.
 */
@Composable
fun SignupScreen(
    session: AppSession,
    phone: String = "",
    onBack: () -> Unit,
    onContinue: (SignupDraft) -> Unit,
) {
    val haptics = LocalVoiidHaptics.current
    val context = LocalContext.current
    var name by remember { mutableStateOf("") }
    var username by remember { mutableStateOf("") }
    var photoJpeg by remember { mutableStateOf<ByteArray?>(null) }
    var photoError by remember { mutableStateOf<String?>(null) }
    var uStatus by remember { mutableStateOf<UStatus>(UStatus.Idle) }

    val nameValid = name.trim().length >= 2
    // FORMAT rules first and locally: 3–20 characters, letters/digits/underscore, no leading
    // digit. Availability is a separate question — see [uStatus]. Mirrors iOS.
    val usernameWellFormed = username.length in 3..20 &&
        username.all { it.isLetterOrDigit() || it == '_' } &&
        !(username.firstOrNull()?.isDigit() == true)
    val canContinue = nameValid && uStatus == UStatus.Available

    // Debounced availability check. An obviously invalid name never costs a round trip.
    LaunchedEffect(username) {
        if (!usernameWellFormed) {
            uStatus = UStatus.Idle
            return@LaunchedEffect
        }
        uStatus = UStatus.Checking
        kotlinx.coroutines.delay(400)   // debounce
        uStatus = try {
            val r = ProfileService(context).checkUsername(username)
            if (r.available) UStatus.Available else UStatus.Taken(r.reason ?: "Username taken")
        } catch (e: Exception) {
            UStatus.Taken("Couldn't check — try again")
        }
    }

    val pickMedia = rememberLauncherForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
        if (uri != null) {
            // The limit is enforced HERE, at pick time — a user with a 12MB photo learns now,
            // not after filling in the rest of the form. Mirrors iOS.
            val bytes = loadAndDownscale(context, uri)
            if (bytes != null) {
                photoJpeg = bytes
                photoError = null
                haptics.success()
            } else {
                photoJpeg = null
                photoError = "That photo couldn't be used. PNG, JPG or WEBP up to 5MB."
                haptics.error()
            }
        }
    }

    OnboardingScaffold(
        footer = {
            OnboardingKitButton(title = "Continue", enabled = canContinue) {
                // Kept in the session too, so anything reading the live profile before the
                // save lands shows the real name rather than a blank.
                session.updateProfile(fullName = name.trim())
                onContinue(
                    SignupDraft(
                        fullName = name.trim(),
                        username = username,
                        photoJpeg = photoJpeg,
                    )
                )
            }
            StepDots(current = 0, total = 2)
        },
    ) {
        Spacer(Modifier.height(VoiidSpacing.md))

        OnboardingHeader(
            title = OnboardingTitleSpec.Stacked("Let's set up", "your profile"),
            blurb = "Add a few details to get started.",
        )

        AvatarPicker(
            photoJpeg = photoJpeg,
            onPick = { pickMedia.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly)) },
            modifier = Modifier.padding(top = VoiidSpacing.md),
        )
        Text(
            if (photoJpeg != null) "Profile photo added" else "Add profile photo",
            style = VoiidFont.rounded(15, FontWeight.SemiBold),
            color = VoiidBrand.text,
            modifier = Modifier.align(Alignment.CenterHorizontally).padding(top = 10.dp),
        )
        Text("PNG, JPG or WEBP. Max 5MB.", style = VoiidFont.rounded(12),
            color = VoiidBrand.textDim,
            modifier = Modifier.align(Alignment.CenterHorizontally).padding(top = 2.dp))
        photoError?.let {
            Text(it, style = VoiidFont.rounded(12), color = VoiidColor.error,
                modifier = Modifier.align(Alignment.CenterHorizontally).padding(top = 4.dp))
        }

        Column(
            Modifier.fillMaxWidth().padding(top = VoiidSpacing.md),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            OnboardingField(
                icon = Icons.Default.Person,
                label = "Full name",
                prompt = "Enter your full name",
                value = name,
                onValueChange = { name = it },
                capitalization = KeyboardCapitalization.Words,
            )

            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                OnboardingField(
                    icon = Icons.Default.AlternateEmail,
                    label = "Username",
                    prompt = "Choose a unique username",
                    value = username,
                    onValueChange = { raw ->
                        // Usernames are lowercase and unspaced. Correcting as they type beats
                        // rejecting after. NO length cap here: iOS shows an over-long entry
                        // and explains it in the help line rather than truncating silently.
                        username = raw.lowercase().filter { it.isLetterOrDigit() || it == '_' }
                    },
                    capitalization = KeyboardCapitalization.None,
                    imeAction = ImeAction.Done,
                    trailing = { UsernameStatusGlyph(uStatus, username, usernameWellFormed) },
                )
                Text(
                    usernameHelp(username, uStatus, usernameWellFormed),
                    style = VoiidFont.rounded(12.5f),
                    // WARNING, not error: a malformed or taken name has not failed anything,
                    // it is a state the user can still type their way out of.
                    color = when {
                        uStatus is UStatus.Available -> VoiidBrand.lime
                        uStatus is UStatus.Taken -> VoiidColor.warning
                        username.isNotEmpty() && !usernameWellFormed -> VoiidColor.warning
                        else -> VoiidBrand.textDim
                    },
                    modifier = Modifier.padding(start = 4.dp),
                )
            }
        }

        VerifiedPhoneRow(phone.ifEmpty { session.profile.phoneNumber })

        Spacer(Modifier.height(VoiidSpacing.xl))
    }
}

/**
 * The trailing glyph inside the username field.
 *
 * A malformed-but-unchecked username shows the WARNING glyph, not the error one: nothing has
 * failed yet, the entry is merely incomplete.
 */
@Composable
private fun UsernameStatusGlyph(status: UStatus, username: String, wellFormed: Boolean) {
    when (status) {
        is UStatus.Idle -> if (username.isNotEmpty() && !wellFormed) {
            Icon(Icons.Outlined.ErrorOutline, null, tint = VoiidColor.warning,
                modifier = Modifier.size(20.dp))
        }
        is UStatus.Checking -> CircularProgressIndicator(
            color = VoiidBrand.textDim, strokeWidth = 2.dp, modifier = Modifier.size(18.dp),
        )
        is UStatus.Available -> Icon(Icons.Outlined.CheckCircle, null, tint = VoiidBrand.lime,
            modifier = Modifier.size(20.dp))
        is UStatus.Taken -> Icon(Icons.Outlined.ErrorOutline, null, tint = VoiidColor.warning,
            modifier = Modifier.size(20.dp))
    }
}

/** The identity half handed from this step to the next (twin of iOS `ProfileDraft`). */
class SignupDraft(
    val fullName: String = "",
    val username: String = "",
    /** Resized JPEG bytes (≤1024px, ≤5MB) ready for upload. */
    val photoJpeg: ByteArray? = null,
)

// MARK: - Avatar

@Composable
private fun AvatarPicker(photoJpeg: ByteArray?, onPick: () -> Unit, modifier: Modifier = Modifier) {
    val avatar = 84.dp
    Box(modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
        Box(contentAlignment = Alignment.BottomEnd) {
            Box(
                modifier = Modifier
                    .size(avatar)
                    .clip(CircleShape)
                    .background(VoiidColor.fieldFill)
                    .border(1.dp, VoiidColor.fieldBorder, CircleShape)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                    ) { onPick() },
                contentAlignment = Alignment.Center,
            ) {
                val bmp = remember(photoJpeg) {
                    photoJpeg?.let { BitmapFactory.decodeByteArray(it, 0, it.size) }
                }
                if (bmp != null) {
                    Image(
                        bitmap = bmp.asImageBitmap(),
                        contentDescription = "Profile photo",
                        contentScale = ContentScale.Crop,
                        modifier = Modifier.size(avatar).clip(CircleShape),
                    )
                } else {
                    Icon(Icons.Default.PhotoCamera, null, tint = VoiidColor.primary,
                         modifier = Modifier.size(26.dp))
                }
            }
            // The affordance stays after a photo exists: it then means "change it".
            Box(
                modifier = Modifier
                    .size(30.dp)
                    .clip(CircleShape)
                    .background(VoiidColor.accent)
                    .border(3.dp, VoiidColor.background, CircleShape)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                    ) { onPick() },
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Default.PhotoCamera, "Pick photo", tint = VoiidColor.primary,
                     modifier = Modifier.size(13.dp))
            }
        }
    }
}

// MARK: - Username

/** Says what is actually true at each stage. Mirrors iOS `usernameHelp`. */
private fun usernameHelp(username: String, status: UStatus, wellFormed: Boolean): String {
    if (username.isEmpty()) return "This will be your unique Voiid ID."
    if (username.length < 3) return "At least 3 characters."
    if (username.first().isDigit()) return "Can't start with a number."
    if (!wellFormed) return "Letters, numbers and underscores only."
    return when (status) {
        is UStatus.Checking -> "Checking availability…"
        is UStatus.Available -> "Available."
        is UStatus.Taken -> status.reason
        is UStatus.Idle -> "This will be your unique Voiid ID."
    }
}

// MARK: - Verified number

/// The number is already proven, so it is shown rather than asked for.
@Composable
private fun VerifiedPhoneRow(phone: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp)
            .padding(top = 16.dp)
            .height(PILL_HEIGHT)
            .clip(RoundedCornerShape(VoiidRadius.pill))
            .background(VoiidColor.bubbleSent)
            .padding(horizontal = 24.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Column(Modifier.weight(1f)) {
            Text("Verified number", style = VoiidFont.rounded(12),
                 color = VoiidColor.textSecondary)
            Text(phone.ifEmpty { "—" }, style = VoiidFont.rounded(16),
                 color = VoiidColor.textPrimary)
        }
        Icon(Icons.Default.CheckCircle, "Verified", tint = VoiidColor.success,
             modifier = Modifier.width(24.dp))
    }
}

// MARK: - Photo loading

/**
 * Reads the picked image, downscales to a 1024px max dimension, compresses JPEG q0.85, and
 * enforces the stated 5MB limit. Returns null when the result would be unusable.
 */
private fun loadAndDownscale(context: Context, uri: Uri): ByteArray? = runCatching {
    val resolver = context.contentResolver
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    resolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it, null, bounds) }
        ?: return null
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) return null

    val maxDimension = 1024
    var sample = 1
    while (maxOf(bounds.outWidth, bounds.outHeight) / (sample * 2) >= maxDimension) sample *= 2
    val opts = BitmapFactory.Options().apply { inSampleSize = sample }
    val decoded = resolver.openInputStream(uri)?.use {
        BitmapFactory.decodeStream(it, null, opts)
    } ?: return null

    val scaled = if (maxOf(decoded.width, decoded.height) > maxDimension) {
        val scale = maxDimension.toFloat() / maxOf(decoded.width, decoded.height)
        Bitmap.createScaledBitmap(
            decoded,
            (decoded.width * scale).toInt().coerceAtLeast(1),
            (decoded.height * scale).toInt().coerceAtLeast(1),
            true,
        )
    } else decoded

    val out = ByteArrayOutputStream()
    scaled.compress(Bitmap.CompressFormat.JPEG, 85, out)
    val bytes = out.toByteArray()
    if (bytes.size > 5 * 1024 * 1024) null else bytes
}.getOrNull()

File: apps/ios/Voiid/Voiid/Onboarding/SignupScreen.swift
//
//  SignupScreen.swift
//  Voiid
//
//  Let's set up your profile — step 1 of 2, shown after verification when the number has NO
//  existing account.
//
//  Built to the design source (`Voiid Ui/Screens/ProfileSetupScreen.swift`) through
//  `OnboardingKit`.
//
//  ── WHY THE PROFILE IS TWO PAGES ────────────────────────────────────────────────
//  Four fields plus an avatar is a wall of form immediately after the user has already done two
//  steps, and the two halves are not the same kind of ask:
//
//    * THIS page is IDENTITY: photo, name, username. Required (bar the photo), and the things
//      that make the account theirs.
//    * The NEXT page is EXTRAS: email and bio. Both optional, and therefore skippable in one
//      tap — which is only honest if they are not mixed in with the required fields.
//
//  Splitting on that seam is what makes page 2 skippable at all. Splitting merely by COUNT would
//  have produced two pages that each have to be completed, which is worse than one.
//
//  ── THE USERNAME CHECK IS REAL ──────────────────────────────────────────────────
//  The design source validates FORMAT only and says so, because it has no server. This screen
//  keeps the app's existing DEBOUNCED AVAILABILITY CHECK against `ProfileService.checkUsername`,
//  so the tick means "this name is free", not merely "this name is well formed". That is a
//  stronger promise than the reference could make and it is worth keeping.
//
//  Nothing is SAVED here — the profile is written once, on the next page, so a user who backs
//  out has not half-created an account. The photo is optional and stays optional: a required
//  avatar at signup is a wall in front of an account the user has already proven they own.
//

import SwiftUI
import PhotosUI

/// What the two profile pages collect between them.
///
/// Hashable as well as Equatable: it travels inside a navigation Route, and NavigationPath
/// requires its values to hash.
struct ProfileDraft: Equatable, Hashable {
    var fullName = ""
    var username = ""
    var email = ""
    var bio = ""
    /// Encoded image data, if the user picked one. Data rather than a `UIImage` so the caller can
    /// hand it straight to an upload without a re-encode.
    var photo: Data?
}

struct SignupScreen: View {
    /// Hands the identity half forward. The caller collects the extras and saves once.
    let onContinue: (ProfileDraft) -> Void
    /// The real verified phone in E.164 (e.g. "+9199..."), shown read-only.
    var phone: String = ""

    @EnvironmentObject var session: AppSession

    @State private var draft = ProfileDraft()
    @State private var pickerItem: PhotosPickerItem?
    @State private var photo: Image?

    /// Username availability, checked against the server as the user types.
    enum UStatus: Equatable { case idle, checking, available, taken(String) }
    @State private var uStatus: UStatus = .idle
    @State private var checkTask: Task<Void, Never>?

    // MARK: Validation

    /// Trimmed, so a name of three spaces does not pass.
    private var nameValid: Bool {
        draft.fullName.trimmingCharacters(in: .whitespaces).count >= 2
    }

    /// FORMAT rules: 3–20 characters, letters/digits/underscore, not starting with a digit.
    /// Availability is a separate question — see `uStatus`.
    private var usernameWellFormed: Bool {
        let u = draft.username
        guard (3...20).contains(u.count) else { return false }
        guard u.allSatisfy({ $0.isLetter || $0.isNumber || $0 == "_" }) else { return false }
        return !(u.first?.isNumber ?? true)
    }

    private var canContinue: Bool { nameValid && uStatus == .available }

    var body: some View {
        ZStack {
            VoiidBrand.ground.ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    OnboardingHeader(
                        title: .stacked("Let's set up", accent: "your profile"),
                        blurb: "Add a few details to get started."
                    )

                    avatarPicker
                        .padding(.top, VoiidSpacing.md)

                    fields
                        .padding(.top, VoiidSpacing.md)

                    verifiedNumber
                        .padding(.top, 12)
                }
                .padding(.horizontal, VoiidSpacing.lg)
                .padding(.bottom, 185)
            }
            .scrollIndicators(.hidden)
            .scrollDismissesKeyboard(.interactively)

            VStack(spacing: 0) {
                Spacer(minLength: 0)
                footer
            }
            .ignoresSafeArea(edges: .bottom)
        }
        .preferredColorScheme(.dark)
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .onChange(of: pickerItem) { _, item in
            Task { await load(item) }
        }
    }

    // MARK: Avatar

    private var avatarPicker: some View {
        VStack(spacing: VoiidSpacing.sm) {
            PhotosPicker(selection: $pickerItem, matching: .images, photoLibrary: .shared()) {
                ZStack(alignment: .bottomTrailing) {
                    Circle()
                        .fill(VoiidBrand.card)
                        .frame(width: 84, height: 84)
                        .overlay(Circle().stroke(VoiidBrand.hairline, lineWidth: 1))
                        .overlay {
                            if let photo {
                                photo
                                    .resizable()
                                    .scaledToFill()
                                    .frame(width: 84, height: 84)
                                    .clipShape(Circle())
                            } else {
                                Image(systemName: "camera")
                                    .font(.system(size: 24, weight: .light))
                                    .foregroundColor(VoiidBrand.lime)
                            }
                        }

                    // The affordance. Present even once a photo is chosen, because it then means
                    // "change it" — and a filled avatar with no visible way to alter it is a
                    // dead end the user has to guess their way out of.
                    Circle()
                        .fill(VoiidBrand.lime)
                        .frame(width: 28, height: 28)
                        .overlay {
                            Image(systemName: photo == nil ? "plus" : "pencil")
                                .font(.system(size: 13, weight: .bold))
                                .foregroundColor(VoiidBrand.onLime)
                        }
                        .overlay(Circle().stroke(VoiidBrand.ground, lineWidth: 3))
                        .offset(x: 2, y: 2)
                }
            }
            .buttonStyle(PressableButtonStyle())
            .accessibilityLabel(photo == nil ? "Add profile photo" : "Change profile photo")

            Text(photo == nil ? "Add profile photo" : "Profile photo added")
                .font(VoiidFont.rounded(15, .semibold))
                .foregroundColor(VoiidBrand.text)

            Text("PNG, JPG or WEBP. Max 5MB.")
                .font(VoiidFont.rounded(12.5))
                .foregroundColor(VoiidBrand.textDim)
        }
        .frame(maxWidth: .infinity)
    }

    /// Loads the picked image and enforces the size limit stated above the picker.
    ///
    /// The limit is checked HERE rather than at upload, because a user who picks a 12MB photo
    /// should learn that immediately, not after filling in the rest of the form.
    private func load(_ item: PhotosPickerItem?) async {
        guard let item,
              let data = try? await item.loadTransferable(type: Data.self),
              let ui = UIImage(data: data)
        else { return }

        let resized = ui.resized(maxDimension: 1024)
        guard let jpeg = resized.jpegData(compressionQuality: 0.85),
              jpeg.count <= 5 * 1024 * 1024
        else {
            Haptics.error()
            return
        }

        draft.photo = jpeg
        photo = Image(uiImage: resized)
        Haptics.success()
    }

    // MARK: Fields

    private var fields: some View {
        VStack(spacing: 12) {
            OnboardingField(
                icon: "person",
                label: "Full name",
                prompt: "Enter your full name",
                text: $draft.fullName,
                contentType: .name,
                autocapitalization: .words
            )

            VStack(alignment: .leading, spacing: 6) {
                OnboardingField(
                    icon: "at",
                    label: "Username",
                    prompt: "Choose a unique username",
                    text: $draft.username,
                    contentType: .username,
                    autocapitalization: .never
                ) {
                    usernameStatusGlyph
                }
                .onChange(of: draft.username) { _, new in onUsernameChange(new) }
                .animation(.easeOut(duration: 0.15), value: uStatus)

                Text(usernameHelp)
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(usernameHelpColor)
                    .padding(.leading, 4)
            }
        }
    }

    /// Only shown once something is typed: a validity marker on an untouched field is noise,
    /// and a red cross before the user has done anything is worse.
    @ViewBuilder
    private var usernameStatusGlyph: some View {
        switch uStatus {
        case .idle:
            if !draft.username.isEmpty && !usernameWellFormed {
                Image(systemName: "exclamationmark.circle")
                    .font(.system(size: 20, weight: .medium))
                    .foregroundColor(VoiidColor.warning)
            }
        case .checking:
            ProgressView().controlSize(.small).tint(VoiidBrand.textDim)
        case .available:
            Image(systemName: "checkmark.circle")
                .font(.system(size: 20, weight: .medium))
                .foregroundColor(VoiidBrand.lime)
        case .taken:
            Image(systemName: "exclamationmark.circle")
                .font(.system(size: 20, weight: .medium))
                .foregroundColor(VoiidColor.warning)
        }
    }

    /// Debounced availability check as the user types.
    ///
    /// The format rules are applied FIRST and locally, so an obviously invalid name never costs
    /// a round trip — and the server is only asked about names it could plausibly accept.
    private func onUsernameChange(_ raw: String) {
        // Usernames are lowercase and unspaced. Correcting as they type beats rejecting after.
        let v = raw.lowercased().filter { $0.isLetter || $0.isNumber || $0 == "_" }
        if v != draft.username { draft.username = v }

        checkTask?.cancel()
        guard usernameWellFormed else { uStatus = .idle; return }
        uStatus = .checking
        checkTask = Task {
            try? await Task.sleep(nanoseconds: 400_000_000)   // debounce
            if Task.isCancelled { return }
            do {
                let r = try await ProfileService.shared.checkUsername(v)
                if Task.isCancelled || v != draft.username { return }
                uStatus = r.available ? .available : .taken(r.reason ?? "Username taken")
            } catch {
                uStatus = .taken("Couldn't check — try again")
            }
        }
    }

    /// Says what is actually true at each stage.
    private var usernameHelp: String {
        if draft.username.isEmpty { return "This will be your unique Voiid ID." }
        if draft.username.count < 3 { return "At least 3 characters." }
        if draft.username.first?.isNumber == true { return "Can't start with a number." }
        if !usernameWellFormed { return "Letters, numbers and underscores only." }
        switch uStatus {
        case .checking:          return "Checking availability…"
        case .available:         return "Available."
        case .taken(let reason): return reason
        case .idle:              return "This will be your unique Voiid ID."
        }
    }

    private var usernameHelpColor: Color {
        if draft.username.isEmpty { return VoiidBrand.textDim }
        switch uStatus {
        case .available: return VoiidBrand.lime
        case .taken:     return VoiidColor.warning
        default:         return usernameWellFormed ? VoiidBrand.textDim : VoiidColor.warning
        }
    }

    // MARK: Verified number

    /// The number is already proven, so it is shown rather than asked for — inert fill, no
    /// border, and a tick. Nothing here is editable.
    private var verifiedNumber: some View {
        HStack(spacing: VoiidSpacing.md) {
            Circle()
                .fill(VoiidBrand.lime.opacity(0.10))
                .frame(width: 38, height: 38)
                .overlay {
                    Image(systemName: "phone")
                        .font(.system(size: 16, weight: .medium))
                        .foregroundColor(VoiidBrand.lime)
                }

            VStack(alignment: .leading, spacing: 2) {
                Text("Verified number")
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(VoiidBrand.textDim)
                Text(phone.isEmpty ? session.profile.phoneNumber : phone)
                    .font(VoiidFont.rounded(16))
                    .foregroundColor(VoiidBrand.text)
            }

            Spacer(minLength: 0)

            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 20))
                .foregroundColor(VoiidColor.success)
        }
        .padding(.horizontal, VoiidSpacing.md)
        .padding(.vertical, 11)
        .background(VoiidBrand.card)
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .stroke(VoiidBrand.hairline, lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
    }

    // MARK: Footer

    private var footer: some View {
        OnboardingFooter {
            OnboardingKitButton(title: "Continue", enabled: canContinue) {
                var d = draft
                d.fullName = draft.fullName.trimmingCharacters(in: .whitespaces)
                // Kept in the session too, so anything reading the live profile before the save
                // lands (the tab bar's avatar, for one) shows the real name rather than a blank.
                session.profile.fullName = d.fullName
                onContinue(d)
            }

            StepDots(current: 0, total: 2)
        }
    }
}
